atef/widgets/ophyd.py

```diff
@@ -0,0 +1,729 @@
+"""
+Ophyd Device-related widgets.
+"""
+from __future__ import annotations
+
+import dataclasses
+import enum
+import logging
+import threading
+import time
+from typing import Any, ClassVar, Dict, List, Optional, Set
+
+import numpy as np
+import ophyd
+import ophyd.device
+from qtpy import QtCore, QtGui, QtWidgets
+from qtpy.QtCore import Qt
+
+from ..qt_helpers import copy_to_clipboard
+from .core import DesignerDisplay
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class OphydAttributeData:
+    attr: str
+    description: Dict[str, Any]
+    docstring: Optional[str]
+    pvname: str
+    read_only: bool
+    readback: Any
+    setpoint: Any
+    signal: ophyd.Signal
+
+    @classmethod
+    def from_device_attribute(
+        cls, device: ophyd.Device, attr: str
+    ) -> OphydAttributeData:
+        """Get attribute information given a device and dotted attribute name."""
+        inst = getattr(device, attr)
+        cpt = getattr(type(device), attr)
+        read_only = isinstance(inst, (ophyd.EpicsSignalRO,))
+        return cls(
+            attr=attr,
+            description={},
+            docstring=cpt.doc,
+            pvname=getattr(inst, "pvname", "(Python)"),
+            read_only=read_only,
+            readback=None,
+            setpoint=None,
+            signal=inst,
+        )
+
+    @classmethod
+    def from_device(cls, device: ophyd.Device) -> Dict[str, OphydAttributeData]:
+        """Create a data dictionary for a given device."""
+        data = {
+            attr: cls.from_device_attribute(device, attr)
+            for attr in device.component_names
+            if attr not in device._sub_devices
+        }
+
+        for sub_name in device._sub_devices:
+            sub_dev = getattr(device, sub_name)
+            sub_data = {
+                f"{sub_name}.{key}": value
+                for key, value in OphydAttributeData.from_device(sub_dev).items()
+            }
+            for key, value in sub_data.items():
+                value.attr = key
+            data.update(sub_data)
+
+        return data
+
+
+class DeviceColumn(enum.IntEnum):
+    """Column information for the device model."""
+    attribute = 0
+    readback = 1
+    setpoint = 2
+    pvname = 3
+
+    total_columns = 4
+
+
+class _DevicePollThread(QtCore.QThread):
+    """
+    Polling thread for updating a PolledDeviceModel.
+
+    Emits ``data_changed(attr: str)`` when an attribute has new data.
+
+    Parameters
+    ----------
+    device : ophyd.Device
+        The ophyd device to poll.
+
+    poll_rate : float
+        The poll rate in seconds. A zero or negative poll rate will indicate
+        single-shot mode.  In "single shot" mode, the data is queried exactly
+        once and then the thread exits.
```

Comment:
I think this is a good decision (poll rate > monitor, allow single shot). Too many devices have fast-updating PVs and it's not the purpose of this GUI to be a live control system viewer.